### ISSUE TITLE
[regression]: a flow of picking up pre-calculated io_properties values is very unclear and bogus

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#10427]
*Installation details*
Scylla version 4.6+:

*Hardware details*
Every "standard" setups.
Today these are as follows:
AWS: i2, i3, i3en
GCP: all instances that have a proper CPU-to-RAM ratio (e.g. n1's n2's, etc.) and have NVMes attached.

*Description*
Before this PR https://github.com/scylladb/scylla/pull/9957 if one runs `scylla_io_setup` on one of the "standard" setups it would not run `iotune` but rather use pre-calculated values to put in io_properties.yaml.

One of the most important aspects of this behavior was, among other things, to avoid statistical fluctuations in the results of `iotune` on the most popular setups used by our users and mimic the regular recommendations of our Field teams for homogeneous cluster setups - run `iotune` on one of the machines a few times, then pick the values that makes the most sense and then use it on the rest of the machines of your cluster with the same HW setup.

Having different configuration in io_properties.yaml is one of main contributors of performance issues since it effectively creates a bottleneck in the cluster. Therefore this should be avoided by all costs.

So, the PR above would effectively deprive everybody why has to install Scylla packages (and not use AMIs for instance) on a "standard" setup from enjoying a homogeneous scylla configuration on all his/her nodes due to `iotune` nature out of the box because `scylla-machine-image` is not a dependency of any `scylla` package, hence is not going to be installed automatically.

More than that, even if it is installed manually then as far as I understand the following running of `scylla_setup` is going to destroy whatever `scylla-machine-image` is going to write in `io_properties.yaml`. 

The bottom line, it's not clear how new and even more importantly *existing* Scylla users are supposed to change their automations in order to continue enjoying pre-calculated values in `io_properties.yaml` after the PR mentioned above.

**Comment by vladzcloudius** [comment]
@avikivity @tarzanek @eliransin FYI

**Comment by eliransin** [comment]
@vladzcloudius I agree that different behaviors for AMI and non AMI setups are bad, however, there are still some things I don't understand. 

> One of the most important aspects of this behavior was, among other things, to avoid statistical fluctuations in the results of `iotune`

How do we tell apart statistical fluctuations from actual performance fluctuations or variations? It sounds to me like the real 
issue is us not trusting `iotune` and not "statistical fluctuations".  We could easily embed the "run iotune a few times" 
recommendation as code into the `iotune` logic. We could have also implement some kind of io configuration propagation
throughout the cluster as part of Scylla if homogeneity of configuration was beyond any doubt crucial gossip or whatever but we 
haven't done that also.

> Having different configuration in io_properties.yaml is one of main contributors of performance issues since it effectively creates a bottleneck in the cluster. Therefore this should be avoided by all costs.

Why do we think that the difference in the properties files is the reason for the bottleneck and not the fact that we might
actually have an instance which falls behind in the sense of io performance. 
Speaking of which, it sounds to me that (assuming quorum writes and not "ALL" for example) we've better work on our load
distribution algorithm for io bound operations instead of trying to artificially create homogeneity in the the cluster - something 
along the lines of HWLB maybe where the parameter for consideration in picking someone up is the percentage of occupation
of io bandwidth or something similar?

> So, the PR above would effectively deprive everybody why has to install Scylla packages (and not use AMIs for instance) on a "standard" setup from enjoying a homogeneous scylla configuration on all his/her nodes due to `iotune` nature out of the box

I know I am probably repeating myself but  why is homogeneous scylla configuration is important instead of 
having homogeneous nodes?

I will just add a little bit of background, recently , we or at least I have a suspicion that AWS instances are not really identical 
and have large variations in the overall performance. One example of this variation is a variation of the wear level of the SSDs.
This is a hypothesis that I appear to share with QA which have started to measure the io performance  properties of the 
instances they use in the prologue part of their performance sensitive tests.
I even heard a claim not long ago that maybe we should do some kind of an `iotune` every once in a while and not only on 
initial startup because there are thing that degrade or change the IO performance over time.
Another claim is that if we would have run `iotune` on cloud deployments it would have shielded us from starting clusters that 
uses faulted instances.

The conclustion of this is that it is not such a bad idea to have users running `iotune` as part of initial scylla setup, they might 
even have a more accurate setup than our current AMIs or cloud deployments. If I missed something and homogeneity of 
configuration is more important than it's accuracy, we can document the process, it seams that the knowledge is there and it is just a matter of putting it in writing. We can also document our measured properties for the different instance types and this
will answer people about how to change the automation (Have the automation create the properties file from known values instead of relying on Scylla installation to do that).

**Comment by tarzanek** [comment]
just my .2 cents
Cloud providers give certain SLAs on their HW
most of hardcoded io properties use these values instead of measured by iotune
(they are generally a bit worse than what we measure)
values from iotune are used where cloud provider didn't set an SLA

above is definitely true for GCP and Azure (I wrote both and invested lots of time into researching and measuring it)
so those providers should be safe

at the same time I agree we should review those values from time to time, GCP did release new SLAs and recommendations recently

so finding a proper middle ground is hard, but playing it safe is what we do now for instances/configs we know;
if everywhere else we go on max values from iotune, then the users/customers should be aware of the risks you mentioned Eliran (like ssd wear)

**Comment by vladzcloudius** [comment]
> I know I am probably repeating myself but why is homogeneous scylla configuration is important instead of having homogeneous nodes?

@eliransin No two computers are ever exactly the same even if they use the same type of HW. For this reason exactly you will never be able to achieve a really homogeneous clusters if you use the actual maximum capabilities of your VMs.

Therefore the right thing to do is to use the configuration values (in io_properties.yaml) low enough that would match the weakest VM provided the HW capabilities fluctuations mentioned above - and use them on all your nodes. 

And this is *exactly* what pre-calculated values are.

Please, let me know if this answers your question?
